src/charm.py

```diff
@@ -6,20 +6,293 @@
 
 logger = logging.getLogger(__name__)
 
+EXTRA_USER_ROLES = "SUPERUSER"
+
 
 class HydraCharm(CharmBase):
     """Charmed Ory Hydra."""
 
     def __init__(self, *args):
         super().__init__(*args)
 
+        self._container_name = "hydra"
+        self._container = self.unit.get_container(self._container_name)
+        self._hydra_config_path = "/etc/config/hydra.yaml"
+        self._name = self.model.app.name
+        self._namespace = self.model.name
+        self._context = {"namespace": self._namespace, "name": self._name}
+
+        database_name = f"{self.app.name.replace('-', '_')}_pg_database"
+        self.pg_database = DatabaseRequires(self, "pg-database", database_name, EXTRA_USER_ROLES)
+
+        self.lightkube_client = Client(namespace=self._namespace, field_manager="lightkube")
+
+        self.resource_handler = KubernetesResourceHandler(
+            template_files=glob.glob("src/manifests/svc.yaml"),
+            context=self._context,
+            field_manager=self._name,
+        )
+
+        for event in [
+            self.on.install,
+            self.on.leader_elected,
+            self.on.upgrade_charm,
+            self.on.config_changed,
+            self.on.hydra_pebble_ready,
+            self.on["pg-database"].relation_changed,
+        ]:
+            self.framework.observe(event, self.main)
+
+        for db_event in [
+            self.pg_database.on.database_created,
+            self.pg_database.on.endpoints_changed,
+        ]:
+            self.framework.observe(db_event, self._apply_secret)
+
+    @property
+    def _hydra_layer(self) -> Layer:
+        """Returns a pre-configured Pebble layer."""
+        layer_config = {
+            "summary": "hydra-operator layer",
+            "description": "pebble config layer for hydra-operator",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "entrypoint of the hydra-operator image",
+                    "command": f"hydra serve all --config {self._hydra_config_path} --dangerous-force-http",
+                    "startup": "enabled",
+                    "environment": {
+                        "SECRETS_SYSTEM": self.model.config["system-secret"],
+                        "SECRETS_COOKIE": self.model.config["cookie-secret"],
+                    },
+                }
+            },
+        }
+        return Layer(layer_config)
+
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer if changed."""
+        try:
+            self._check_container_connection()
+        except CheckFailedError as err:
+            self.model.unit.status = err.status
+            return
+
+        self._push_config()
+
+        try:
+            self.resource_handler.apply()
+            self.lightkube_client.patch(
+                StatefulSet,
+                self._name,
+                {
```

Comment:
Why is this a python object instead of a jinja template?

Reply:
Since this was an initial deployment and the main objective was to make hydra work, I'm patching the statefulset created by pebble with the secret data created on `database_created` event. So for now it's just a workaround.
I think we need a discussion on how to fetch and store the relation data, so I will handle it in another PR.